### ISSUE TITLE
Unable to use local database

### ISSUE THREAD

**Issue by ekrell** [kaspermunch/sap#2]
I am attempting to use sap with a local NCBI-formatted FASTA database. 
Thanks in advance for any assistance. 

Version: 
`sap 1.9.7`

Command:
`sap --project mine -x 90 test.fasta --database 16SMicrobial.fasta`

test.fasta: https://pastebin.com/1FUkYZn8
first entry of 16SMicrobial.fasta: https://pastebin.com/eLgf0k88

Output: 
```
Checking cache for deprecated entries

Building native SAP database... 
## SAP crashed, sorry ###################################################
Help creating a more stable program by sending all the debugging information
between the lines and your SAP version number to kaspermunch@gmail.com along
with *.sap file in the project folder and the sequence input file used.

  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/ConsoleScripts.py", line 220, in sap
    homolcompiler = HomolCompiler(options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Homology.py", line 108, in __init__
    self.db = plugin.DB(self.options.database, self.options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 61, in __init__
    self._build()
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 123, in _build
    raise WrongFormatError(fastaRecord.title)


Wrong format of FASTA header line: gi|645322056|ref|NR_118889.1| Amycolatopsis azurea strain NRRL 11412 16S ribosomal RNA gene, partial sequence
#########################################################################
```

To reduce problem complexity, I made a version of the database FASTA that had only GI numbers

Command: 
`sap --project mine -x 90 test.fasta --database 16SMicrobial_simple.fasta`

first entry of 16SMicrobial_simple.fasta: https://pastebin.com/jTsZKdA8

Output:
```
Checking cache for deprecated entries

Building native SAP database... 
## SAP crashed, sorry ###################################################
Help creating a more stable program by sending all the debugging information
between the lines and your SAP version number to kaspermunch@gmail.com along
with *.sap file in the project folder and the sequence input file used.

  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/ConsoleScripts.py", line 220, in sap
    homolcompiler = HomolCompiler(options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Homology.py", line 108, in __init__
    self.db = plugin.DB(self.options.database, self.options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 61, in __init__
    self._build()
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 123, in _build
    raise WrongFormatError(fastaRecord.title)


Wrong format of FASTA header line: 645322056
#########################################################################
```

mine.sap: https://pastebin.com/JRxRjFKd

**Comment by kaspermunch** [comment]
Hi Evan,

You local database needs to be formatted in a special way for SAP to read it. See the info pasted in below. However, if the sequences you want in your local database are in GenBank it is much easier/faster to have SAP compile the database for you using the the

Compiling a local database specified by a query for Genbank nucleotide db:
sap --compile '(COI[Gene Name]) AND barcode[Keyword]' --database coi_barcode_db.fasta

Running against a local database:
sap --project myproject --database coi_barcode_db.fasta query.fasta

Hope this helps,
Kasper



Manually compiling a local database:

The database file needs to be in Fasta format. The taxonomic annotation for each sequence can be supplied either by using the NCBI formatted header line like this:


>gi|222142915|gb|FJ606797.1| Bombus ignitus vitellogenin mRNA, complete cds
AGCGATAGGATCACGTAGACCACAGATATAGCGATAGGATCACGTAGACCACAGATAT
TAGGATCACGTAGACCACAGATATAGCGATAGGATCACGTAGACCACAGATATCCTAA


or just the GI number as sequence id like this:


>222142915
AGCGATAGGATCACGTAGACCACAGATATAGCGATAGGATCACGTAGACCACAGATAT
TAGGATCACGTAGACCACAGATATAGCGATAGGATCACGTAGACCACAGATATCCTAA


in which case the taxonomic annotation is retrieved from NCBI. You can also supply the taxonomic annotation in the header line in following format. Note the user of colons and semicolons:


>mySequenceID ; family: Hominidae, genus: Homo, species: Homo sapiens ; Myself
AGCGATAGGATCACGTAGACCACAGATATAGCGATAGGATCACGTAGACCACAGATAT
TAGGATCACGTAGACCACAGATATAGCGATAGGATCACGTAGACCACAGATATCCTAA


The following taxonomic levels are recognized: superkingdom, kingdom, subkingdom, superphylum, phylum, subphylum, superclass, class, subclass, infraclass, superorder, order, suborder, infraorder, parvorder, superfamily, family, subfamily, supertribe, tribe, subtribe, supergenus, genus, subgenus, species group, species subgroup, species, subspecies. The name after the last semicolon is the organism name. This is sometimes different from the species or subspecies name.



On 11 May 2017, 20.30 +0200, Evan Krell <notifications@github.com>, wrote:

I am attempting to use sap with a local NCBI-formatted FASTA database.
Thanks in advance for any assistance.

Version:
sap 1.9.7

Command:
sap --project mine -x 90 test.fasta --database 16SMicrobial.fasta

test.fasta: https://pastebin.com/1FUkYZn8
first entry of 16SMicrobial.fasta: https://pastebin.com/eLgf0k88

Output:


Checking cache for deprecated entries

Building native SAP database...
## SAP crashed, sorry ###################################################
Help creating a more stable program by sending all the debugging information
between the lines and your SAP version number to kaspermunch@gmail.com along
with *.sap file in the project folder and the sequence input file used.

  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/ConsoleScripts.py", line 220, in sap
    homolcompiler = HomolCompiler(options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Homology.py", line 108, in __init__
    self.db = plugin.DB(self.options.database, self.options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 61, in __init__
    self._build()
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 123, in _build
    raise WrongFormatError(fastaRecord.title)


Wrong format of FASTA header line: gi|645322056|ref|NR_118889.1| Amycolatopsis azurea strain NRRL 11412 16S ribosomal RNA gene, partial sequence
#########################################################################



To reduce problem complexity, I made a version of the database FASTA that had only GI numbers

Command:
sap --project mine -x 90 test.fasta --database 16SMicrobial_simple.fasta

first entry of 16SMicrobial_simple.fasta: https://pastebin.com/jTsZKdA8

Output:


Checking cache for deprecated entries

Building native SAP database...
## SAP crashed, sorry ###################################################
Help creating a more stable program by sending all the debugging information
between the lines and your SAP version number to kaspermunch@gmail.com along
with *.sap file in the project folder and the sequence input file used.

  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/ConsoleScripts.py", line 220, in sap
    homolcompiler = HomolCompiler(options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Homology.py", line 108, in __init__
    self.db = plugin.DB(self.options.database, self.options)
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 61, in __init__
    self._build()
  File "/usr/local/lib/python2.7/dist-packages/SAP-1.9.7-py2.7-linux-x86_64.egg/SAP/Databases/Native.py", line 123, in _build
    raise WrongFormatError(fastaRecord.title)


Wrong format of FASTA header line: 645322056
#########################################################################



mine.sap: https://pastebin.com/JRxRjFKd

**Comment by ekrell** [comment]
Thank you for your help. 

What I have tried is the first two options "NCBI-formatted" and "GI only".
I understand from the documentation that it should be able to query NCBI for the taxonomic information, but it simply crashes if I try. 

Looking at the code and crash report, it happens here:

```
     for fastaRecord in fastaIterator:

            taxonomy = Taxonomy.Taxonomy()
            try:
                identifier, taxonomyString, organismName = re.split(r'\s*;\s*', fastaRecord.title.strip())
                taxonomy.populateFromString(taxonomyString)
                taxonomy.organism = organismName
            except (ValueError, Taxonomy.ParsingError) as exc:
                raise WrongFormatError(fastaRecord.title)
            if not (identifier and taxonomyString and organismName):
raise WrongFormatError(fastaRecord.title)
```
Even though I only have a GI, it believes that I am using the native SAP format, and attempts to get the GI, Taxon and name. 

I know that I can simply build a valid database from within SAP, but I need to use specific GIs. 
This is because I am comparing various assignment methods. The original database used in Blast's preformatted nucleotide database which has tons of stuff. I can check to see if Genbank has them all. If so, then I can filter the database created by SAP to have just the ones I need. But I suspect that Blast has more. 

Last Friday I wrote a script to build an SAP-formatted Fasta from a list of GIs. It ran all weekend then finally crashed. Perhaps NCBI kicked me out. I did not investigate because it was taking too long anyway. Today I intend to modify it to use a local copy of the NCBI taxonomy database, which should tremendously speed up building SAP-formatted FASTAs.

**Comment by kaspermunch** [comment]
Yes, if you just specify GI SAP will get the taxonomy for you from the database. It needs that information to make the analysis. I would just use SAP to compile the database and then remove the GIs from the local database that you don’t want in there.

Cheers,
Kasper

On 30 May 2017, 18.45 +0200, Evan Krell <notifications@github.com>, wrote:

Thank you for your help.

What I have tried is the first two options "NCBI-formatted" and "GI only".
I understand from the documentation that it should be able to query NCBI for the taxonomic information, but it simply crashes if I try.

Looking at the code and crash report, it happens here:


     for fastaRecord in fastaIterator:

            taxonomy = Taxonomy.Taxonomy()
            try:
                identifier, taxonomyString, organismName = re.split(r'\s*;\s*', fastaRecord.title.strip())
                taxonomy.populateFromString(taxonomyString)
                taxonomy.organism = organismName
            except (ValueError, Taxonomy.ParsingError) as exc:
                raise WrongFormatError(fastaRecord.title)
            if not (identifier and taxonomyString and organismName):
raise WrongFormatError(fastaRecord.title)



Even though I only have a GI, it believes that I am using the native SAP format, and attempts to get the GI, Taxon and name.

I know that I can simply build a valid database from within SAP, but I need to use specific GIs.
This is because I am comparing various assignment methods. The original database used in Blast's preformatted nucleotide database which has tons of stuff. I can check to see if Genbank has them all. If so, then I can filter the database created by SAP to have just the ones I need. But I suspect that Blast has more.

Last Friday I wrote a script to build an SAP-formatted Fasta from a list of GIs. It ran all weekend then finally crashed. Perhaps NCBI kicked me out. I did not investigate because it was taking too long anyway. Today I intend to modify it to use a local copy of the NCBI taxonomy database, which should tremendously speed up building SAP-formatted FASTAs.
